### ISSUE TITLE
I am truly grateful to the authors for open-sourcing code

### ISSUE THREAD

**Issue by newer7** [dominickrei/Limited-data-vits#2]
I am truly grateful to the authors for open-sourcing code, which introduces a novel approach. However, I seem to have encountered an issue. For datasets like Flowers102, Chaoyang, PMNIST, ClipArt, Infograph, and Sketch, how should the validation set be partitioned? Is it the traditional 8:2 split, or is the validation set the same as the test set? Especially for Flowers102, there is a significant performance gap between using the ViT-T + SSAT method and the results reported in the paper. If you could provide some clarification, I would greatly appreciate it. Once again, thank you for the authors' contributions.

**Comment by srijandas07** [comment]
Hi @newer7 
All our hyperparams are set based on CIFAR and IN1K experiments. Also, we have constant # of epochs to train our models. So, we didn't use a val set for small datasets like Flowes102 (following the other smallViT paper's training protocols).

**Comment by newer7** [comment]
> Hi @newer7 All our hyperparams are set based on CIFAR and IN1K experiments. Also, we have constant # of epochs to train our models. So, we didn't use a val set for small datasets like Flowes102 (following the other smallViT paper's training protocols).

Thank you very much for your reply.

**Comment by newer7** [comment]
> Hi @newer7 All our hyperparams are set based on CIFAR and IN1K experiments. Also, we have constant # of epochs to train our models. So, we didn't use a val set for small datasets like Flowes102 (following the other smallViT paper's training protocols).

Thank you very much. Could you please provide some insights into the implementation details of CVT-SSAT and Swin-SSAT?the decoder design from [20] for ViT, and utilize the decoder design from ConvMAE [16] and SimMIM [59] for hierarchical encoders such as CVT and Swin, respectively. Is the only difference in the design of the decoder?

**Comment by dominickrei** [comment]
Hi @newer7, thank you for your interest. For CVT+SSAT and Swin+SSAT experiments we follow the same encoder-decoder designs proposed in [ConvMAE for CVT](https://github.com/Alpha-VL/ConvMAE), and [SimMIM for Swin](https://github.com/microsoft/SimMIM).

**Comment by aluo7** [comment]
> Hi @newer7 All our hyperparams are set based on CIFAR and IN1K experiments. Also, we have constant # of epochs to train our models. So, we didn't use a val set for small datasets like Flowes102 (following the other smallViT paper's training protocols).

Hi, does this mean you also don't use validation for C10, C100, and SVHN, as they're all indicated as small datasets on the paper? Also, are there any other differences in configurations with experiments for Flowers102 and SVHN in comparison to Table 14 (i.e. no horizontal flip for SVHN)?

**Comment by dominickrei** [comment]
Hi @aluo7, that is correct, we also do not use a validation set on C10, C100, or SVHN.

As for experimental configuration, we exactly follow the protocol illustrated in Table 14 (so we are doing horizontal flipping on SVHN with 0.5 probability). Is there a configuration detail you are interested in that is not listed in Table 14? I'm happy to share further details.

**Comment by aluo7** [comment]
@dominickrei thanks for the response!

I'm curious - did you observe a large variability in accuracy results for SVHN? I'm trying to reproduce your results, but am getting an accuracy anywhere from 0.65-0.96 for this dataset. I noticed you only report Top-1 accuracy, so I was wondering if you encountered this as well. My assumption is that it's due to horizontal flip, and varies largely depending on how the test-set is split.

**Comment by dominickrei** [comment]
Hello, this is an interesting observation and is not one we experienced while conducting our own experiments.

Would you be able to provide some additional details of your training setup? I would like to try and reproduce this behavior. Specifically:
- Which size ViT are you using in your experiments (small, base, etc)?
- How are you implementing the data loading of SVHN? Are you using [`torchvision.datasets.SVHN`](https://pytorch.org/vision/main/generated/torchvision.datasets.SVHN.html) or have you written the dataloading pipeline?
- Which random seeds are you using?
- Have you made any changes to the hyperparameters from the ones in the code?

Thank you!

**Comment by dominickrei** [comment]
I attempted to reproduce this in my own environment with ViT-Tiny and got the following results after 3 runs:
- Run 1 (seed=0): 97.407%
- Run 2 (seed=1234): 97.507%
- Run 3 (seed=5678): 97.515%

Steps I took:
- Clone this repository
- Add the following to line 34 of `util/datasets.py`
```
elif args.data_path == 'svhn':
    split = 'train' if is_train else 'test'
    dataset = datasets.SVHN(root='./data', split=split, download=True, transform=transform)
```
- Command I am using: 
```
CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7 python -m torch.distributed.launch --nproc_per_node=8 --master_port 9999 main_two_branch.py --data_path svhn --output_dir ./output_dir_svhn-run --seed <SEE_ABOVE> --epochs 100 --model mae_vit_tiny --input_size 32 --patch_size 2
```

Is this setting similar to yours? I'm happy to help troubleshoot any possible issues. 

My system is a machine with 8 NVIDIA A5000 GPUs and each training run took a maximum of 8GB of VRAM, training time is ~3 hours. However, the top-1 results should be identical if trained on less than 8 GPUs.

Additionally, ive added SVHN dataloading to the repository, so feel free to do a `git pull`.

**Comment by aluo7** [comment]
Hi @dominickrei, I should specify that these are just for reproducing ViT-T results **without** SSAT on SVHN, using my own implementation. Sorry if I caused any confusion!

> * Which size ViT are you using in your experiments (small, base, etc)?
> 
> * How are you implementing the data loading of SVHN? Are you using [`torchvision.datasets.SVHN`](https://pytorch.org/vision/main/generated/torchvision.datasets.SVHN.html) or have you written the dataloading pipeline?
> 
> * Which random seeds are you using?
> 
> * Have you made any changes to the hyperparameters from the ones in the code?

* Tiny ViT configurations
* I'm using [`torchvision.datasets.SVHN`](https://pytorch.org/vision/main/generated/torchvision.datasets.SVHN.html), correct
* I'm using seed=0
* No, I have not made any changes in the hyperparameters from Table 14 or the default configurations specified in `main_two_branch.py`

Your experiments are reporting top-1 accuracy, correct? I'm wondering if the variability is being observed beyond the top-1 accuracy. I suspect that this might be due to how data augmentation, particularly horizontal flipping, is impacting the labels of the images. This could lead to issues when evaluating on the test set—especially if the split is unfavorable, since data augmentation is being removed during testing. Could it be that the variability I'm observing is more pronounced when looking beyond top-1 accuracy, possibly due to the nature of SVHN or how the test set is split?

**Comment by dominickrei** [comment]
Hi @aluo7, I see!

Just to clarify, are you using a modified version of our repository for training without SSAT? I don't believe our repository supports the standard MAE training protocol that is reported in our paper, we only provide the code for training with SSAT.

The code used to obtain results without SSAT (i.e., SSL+FT) is available in the [MAE codebase](https://github.com/facebookresearch/mae).

You are correct, our experiments report the top-1 accuracy. Which metrics are you using to observe this behavior? In a previous post you mentioned accuracy, but it is my understanding that standard accuracy (`acc = num_correct / num_samples`) is identical to top-1 accuracy.

Can you clarify your concern with horizontal flipping on SVHN? I understand that it represents a distribution of the data that is not seen at test time, but this is the case for most data augmentation techniques (CutMix, MixUp, etc). Additionally, if you remove horizontal flipping from your model, do you still observe the same behavior?

**Comment by aluo7** [comment]
@dominickrei wait, I'm sorry this is confusing me. Are the reported accuracies in table 1 for ViT variants without SSAT actually MAEs? I assumed the description of the encoder/decoder architecture in section 5 was referring to training with SSAT, and that the ViT-T reported accuracies were vanilla ViTs trained from scratch.

**Comment by aluo7** [comment]
Just to follow up:


> Just to clarify, are you using a modified version of our repository for training without SSAT? I don't believe our repository supports the standard MAE training protocol that is reported in our paper, we only provide the code for training with SSAT.

Yes, I am using a modified version. Though I was just training a vanilla ViT-T, so that's likely the delta if you're reporting MAE accuracies from scratch.

> You are correct, our experiments report the top-1 accuracy. Which metrics are you using to observe this behavior? In a previous post you mentioned accuracy, but it is my understanding that standard accuracy (`acc = num_correct / num_samples`) is identical to top-1 accuracy.

You're correct here, I was just wondering if there was anything else that could contribute to it. As you said, I'm using standard accuracy which is equivalent to top-1 accuracy.

> Can you clarify your concern with horizontal flipping on SVHN? I understand that it represents a distribution of the data that is not seen at test time, but this is the case for most data augmentation techniques (CutMix, MixUp, etc). Additionally, if you remove horizontal flipping from your model, do you still observe the same behavior?

My concern is that excessive data augmentation in the form of flips or rotations can frequently end up changing the label of an image (i.e. rotating a 6 can change the label to a 9). For example, see the following on [MNIST](https://datadote.github.io/blog/fastpages/jupyter/2020/04/07/Tutorial_MNIST_Data_Aug.html). When removing horizontal flip, as well as Rand Augment (I'm assuming due to rotations), I'm able to achieve much more stable performance. Though this is with a vanilla ViT rather than an MAE, so it may be a different story from what is reported in your experiments. 

edit: a better example of how a label may be changed with a horizontal switch (again, for mnist): https://www.researchgate.net/figure/Examples-of-data-augmentation-transformations-on-an-image-from-the-MNIST-dataset-a_fig2_346701220

**Comment by dominickrei** [comment]
> wait, I'm sorry this is confusing me. Are the reported accuracies in table 1 for ViT variants without SSAT actually MAEs? I assumed the description of the encoder/decoder architecture in section 5 was referring to training with SSAT, and that the ViT-T reported accuracies were vanilla ViTs trained from scratch.

You are correct that the results in Table 1 without SSAT are the vanilla ViT's. I apologize, I worded my response poorly. What I mean to say is that the [MAE codebase](https://github.com/facebookresearch/mae) supports both SSL+FT training and vanilla ViT training which we do not support in this codebase. So it is not possible to reproduce any of our baseline results (e.g., vanilla ViT, SSL+FT) using this repo without modification.

---

> Yes, I am using a modified version. Though I was just training a vanilla ViT-T, so that's likely the delta if you're reporting MAE accuracies from scratch.

You're right to train a vanilla ViT-T if this is the baseline you're interested in. I incorrectly assumed it was SSL+FT.

**Instead of using a modified version can you try using the [MAE codebase](https://github.com/facebookresearch/mae) to train this?** I ran some training runs using their code overnight and could not reproduce the behavior. Here are my details:

```
CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7 python -m torch.distributed.launch --master_port 9998 --nproc_per_node=8 main_finetune.py --data_path svhn --output_dir ./svhn_runX --seed <same as above> --epochs 100 --model vit_tiny --input_size 32 --patch_size 2
```

| Training run | Top-1 Acc | Top-5 Acc |
|----------|----------|----------|
| 1   | 93.97   | 99.32 |
| 2   | 94.66   | 99.39 |
| 3   | 94.31   | 99.39 |

---

>My concern is that excessive data augmentation in the form of flips or rotations can frequently end up changing the label of an image

I see what you mean (and thank you for providing these supplementary materials), but I still have doubts that it is causing the variability you're seeing.

The first article isnt representative of our setting as they perform horizontal flip with a probability of 1.0, while we are using a probability of 0.5. If I train ViT-T on SVHN with a horizontal flip probability of 1.0 I get a top-1 accuracy of 41.43% (vs 94.31% above). Which makes sense because the model never sees data at the correct orientation during train time.

It is surprising that removing horizontal flipping and RandAugment leads to more **stable** performance over multiple runs since you are using the same seed each run. Also, in my experiments I do not see variability despite using these augmentations. Can you try training ViT-T using the MAE codebase and see if these issues persist?
